File: cli/cmd/devices.go
package cmd

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

var devicesCmd = &cobra.Command{
	Use:   "devices",
	Short: "List and revoke devices on your account",
}

var devicesListCmd = &cobra.Command{
	Use:   "list",
	Short: "List devices registered for your account",
	Args:  cobra.NoArgs,
	RunE:  runDevicesList,
}

var (
	devicesRevokeYes bool
)

var devicesRevokeCmd = &cobra.Command{
	Use:   "revoke <device-id>",
	Short: "Revoke a device (closes any in-flight WebSocket sessions for it)",
	Args:  cobra.ExactArgs(1),
	RunE:  runDevicesRevoke,
}

func init() {
	devicesRevokeCmd.Flags().BoolVar(&devicesRevokeYes, "yes", false,
		"skip the confirmation prompt")

	devicesCmd.AddCommand(devicesListCmd)
	devicesCmd.AddCommand(devicesRevokeCmd)
}

func runDevicesList(_ *cobra.Command, _ []string) error {
	client, _, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	devices, err := client.ListUserDevices()
	if err != nil {
		return err
	}
	if jsonOut {
		out, _ := json.Marshal(devices)
		fmt.Println(string(out))
		return nil
	}
	if len(devices) == 0 {
		fmt.Println("(no devices registered)")
		return nil
	}
	fmt.Printf("%-12s  %-30s  %-20s  %-20s  %s\n", "ID", "LABEL", "CREATED", "LAST SEEN", "STATUS")
	for _, d := range devices {
		last := "(never)"
		if d.LastSeenAt != nil {
			last = d.LastSeenAt.Local().Format(time.RFC3339)
		}
		status := "active"
		if !d.IsActive {
			status = "revoked"
		}
		fmt.Printf("%-12d  %-30s  %-20s  %-20s  %s\n",
			d.DeviceID, d.Label, d.CreatedAt.Local().Format(time.RFC3339), last, status)
	}
	return nil
}

func runDevicesRevoke(_ *cobra.Command, args []string) error {
	id, err := strconv.ParseInt(args[0], 10, 64)
	if err != nil {
		return fmt.Errorf("device-id must be a number, got %q", args[0])
	}

	client, _, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	if !devicesRevokeYes {
		// Destructive: closes any in-flight WS sessions signed by this
		// device's key (server-side WithRevokeHook in main.go).
		fmt.Fprintf(os.Stderr,
			"Revoke device %d? This closes its active sessions. [y/N]: ", id)
		reader := bufio.NewReader(os.Stdin)
		ans, _ := reader.ReadString('\n')
		ans = strings.ToLower(strings.TrimSpace(ans))
		if ans != "y" && ans != "yes" {
			return fmt.Errorf("aborted")
		}
	}

	if err := client.RevokeUserDevice(id); err != nil {
		return err
	}
	if jsonOut {
		fmt.Printf(`{"deviceId":%d,"revoked":true}`+"\n", id)
	} else {
		fmt.Printf("Revoked device %d\n", id)
	}
	return nil
}

File: crates/kutup-cli/src/commands/devices.rs
//! `kutup devices` — list and revoke account devices. Mirrors `cmd/devices.go`.

use anyhow::Result;
use clap::Subcommand;

use crate::commands::confirm;
use crate::context::require_session;

#[derive(Subcommand)]
pub enum DevicesCmd {
    /// List devices registered for your account.
    List,
    /// Revoke a device (closes its in-flight WebSocket sessions).
    Revoke {
        /// Numeric device id.
        device_id: i64,
        /// Skip the confirmation prompt.
        #[arg(long)]
        yes: bool,
    },
}

pub fn run(profile: &str, json: bool, cmd: &DevicesCmd) -> Result<()> {
    match cmd {
        DevicesCmd::List => list(profile, json),
        DevicesCmd::Revoke { device_id, yes } => revoke(profile, json, *device_id, *yes),
    }
}

fn list(profile: &str, json: bool) -> Result<()> {
    let ctx = require_session(profile)?;
    let devices = ctx.client.list_user_devices()?;

    if json {
        crate::output::print_json(&devices)?;
        return Ok(());
    }
    if devices.is_empty() {
        println!("(no devices registered)");
        return Ok(());
    }
    println!(
        "{}",
        crate::output::header(format!(
            "{:<12}  {:<30}  {:<16}  {:<16}  STATUS",
            "ID", "LABEL", "CREATED", "LAST SEEN"
        ))
    );
    for d in &devices {
        let last = d
            .last_seen_at
            .as_deref()
            .map(crate::output::format_time)
            .unwrap_or_else(|| "(never)".to_string());
        let status = if d.is_active { "active" } else { "revoked" };
        println!(
            "{:<12}  {:<30}  {:<16}  {:<16}  {}",
            d.device_id,
            d.label,
            crate::output::format_time(&d.created_at),
            last,
            status
        );
    }
    Ok(())
}

fn revoke(profile: &str, json: bool, device_id: i64, yes: bool) -> Result<()> {
    let ctx = require_session(profile)?;

    confirm(
        &format!("Revoke device {device_id}? This closes its active sessions."),
        yes,
    )?;

    ctx.client.revoke_user_device(device_id)?;
    if json {
        crate::output::print_json(&serde_json::json!({ "deviceId": device_id, "revoked": true }))?;
    } else {
        println!("Revoked device {device_id}");
    }
    Ok(())
}
